// @riotjs/util/index.cjs
'use strict';

var bindingTypes = require('./binding-types.cjs');
var checks = require('./checks.cjs');
var constants = require('./constants.cjs');
var dom = require('./dom.cjs');
var expressionTypes = require('./expression-types.cjs');
var functions = require('./functions.cjs');
var misc = require('./misc.cjs');
var objects = require('./objects.cjs');
var strings = require('./strings.cjs');



exports.EACH = bindingTypes.EACH;
exports.IF = bindingTypes.IF;
exports.SIMPLE = bindingTypes.SIMPLE;
exports.SLOT = bindingTypes.SLOT;
exports.TAG = bindingTypes.TAG;
exports.checkType = checks.checkType;
exports.isBoolean = checks.isBoolean;
exports.isFunction = checks.isFunction;
exports.isNil = checks.isNil;
exports.isNode = checks.isNode;
exports.isObject = checks.isObject;
exports.isSvg = checks.isSvg;
exports.isTemplate = checks.isTemplate;
exports.ATTRIBUTES_KEY_SYMBOL = constants.ATTRIBUTES_KEY_SYMBOL;
exports.COMPONENTS_IMPLEMENTATION_MAP = constants.COMPONENTS_IMPLEMENTATION_MAP;
exports.DOM_COMPONENT_INSTANCE_PROPERTY = constants.DOM_COMPONENT_INSTANCE_PROPERTY;
exports.IS_COMPONENT_UPDATING = constants.IS_COMPONENT_UPDATING;
exports.IS_DIRECTIVE = constants.IS_DIRECTIVE;
exports.IS_PURE_SYMBOL = constants.IS_PURE_SYMBOL;
exports.MOUNT_METHOD_KEY = constants.MOUNT_METHOD_KEY;
exports.ON_BEFORE_MOUNT_KEY = constants.ON_BEFORE_MOUNT_KEY;
exports.ON_BEFORE_UNMOUNT_KEY = constants.ON_BEFORE_UNMOUNT_KEY;
exports.ON_BEFORE_UPDATE_KEY = constants.ON_BEFORE_UPDATE_KEY;
exports.ON_MOUNTED_KEY = constants.ON_MOUNTED_KEY;
exports.ON_UNMOUNTED_KEY = constants.ON_UNMOUNTED_KEY;
exports.ON_UPDATED_KEY = constants.ON_UPDATED_KEY;
exports.PARENT_KEY_SYMBOL = constants.PARENT_KEY_SYMBOL;
exports.PLUGINS_SET = constants.PLUGINS_SET;
exports.PROPS_KEY = constants.PROPS_KEY;
exports.ROOT_KEY = constants.ROOT_KEY;
exports.SHOULD_UPDATE_KEY = constants.SHOULD_UPDATE_KEY;
exports.SLOTS_KEY = constants.SLOTS_KEY;
exports.STATE_KEY = constants.STATE_KEY;
exports.TEMPLATE_KEY_SYMBOL = constants.TEMPLATE_KEY_SYMBOL;
exports.UNMOUNT_METHOD_KEY = constants.UNMOUNT_METHOD_KEY;
exports.UPDATE_METHOD_KEY = constants.UPDATE_METHOD_KEY;
exports.VALUE_ATTRIBUTE = constants.VALUE_ATTRIBUTE;
exports.DOMattributesToObject = dom.DOMattributesToObject;
exports.cleanNode = dom.cleanNode;
exports.clearChildren = dom.clearChildren;
exports.insertBefore = dom.insertBefore;
exports.moveChildren = dom.moveChildren;
exports.removeChild = dom.removeChild;
exports.replaceChild = dom.replaceChild;
exports.ATTRIBUTE = expressionTypes.ATTRIBUTE;
exports.EVENT = expressionTypes.EVENT;
exports.TEXT = expressionTypes.TEXT;
exports.VALUE = expressionTypes.VALUE;
exports.autobindMethods = functions.autobindMethods;
exports.callOrAssign = functions.callOrAssign;
exports.noop = functions.noop;
exports.evaluateAttributeExpressions = misc.evaluateAttributeExpressions;
exports.memoize = misc.memoize;
exports.panic = misc.panic;
exports.cloneDeep = objects.cloneDeep;
exports.defineDefaults = objects.defineDefaults;
exports.defineProperties = objects.defineProperties;
exports.defineProperty = objects.defineProperty;
exports.camelToDashCase = strings.camelToDashCase;
exports.dashToCamelCase = strings.dashToCamelCase;
