// riot/esm/api/__.js
/* Riot v9.1.8, @license MIT */
import bindingTypes from '../dependencies/@riotjs/util/binding-types.js';
import { PROPS_KEY, STATE_KEY, IS_COMPONENT_UPDATING, ATTRIBUTES_KEY_SYMBOL, COMPONENTS_IMPLEMENTATION_MAP, PLUGINS_SET, DOM_COMPONENT_INSTANCE_PROPERTY, PARENT_KEY_SYMBOL } from '../dependencies/@riotjs/util/constants.js';
import expressionTypes from '../dependencies/@riotjs/util/expression-types.js';
import { template as create, createBinding as create$1, createExpression as create$4 } from '../dependencies/@riotjs/dom-bindings/dist/dom-bindings.js';
import cssManager from '../core/css-manager.js';

// expose some internal stuff that might be used from external tools
const __ = {
  cssManager,
  DOMBindings: {
    template: create,
    createBinding: create$1,
    createExpression: create$4,
    bindingTypes,
    expressionTypes,
  },
  globals: {
    PROPS_KEY,
    STATE_KEY,
    IS_COMPONENT_UPDATING,
    ATTRIBUTES_KEY_SYMBOL,
    COMPONENTS_IMPLEMENTATION_MAP,
    PLUGINS_SET,
    DOM_COMPONENT_INSTANCE_PROPERTY,
    PARENT_KEY_SYMBOL,
  },
};

export { __ };
